samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java

```diff
@@ -100,77 +100,75 @@ public void init(Config config, TaskContext context) {
       public void init(Config config, TaskContext context) {
         filterFn.init(config, context);
       }
-    }, output, OperatorSpec.OpCode.FILTER, graph.getNextOpId());
+    }, nextStream, OperatorSpec.OpCode.FILTER, opId);
   }
 
   /**
    * Creates a {@link StreamOperatorSpec}.
    *
    * @param transformFn  the transformation function
-   * @param graph  the {@link StreamGraphImpl} object
-   * @param output  the output {@link MessageStreamImpl} object
+   * @param nextStream  the output {@link MessageStreamImpl} to send messages to
+   * @param opId  the unique ID of the operator
    * @param <M>  type of input message
    * @param <OM>  type of output message
    * @return  the {@link StreamOperatorSpec}
    */
   public static <M, OM> StreamOperatorSpec<M, OM> createStreamOperatorSpec(
-      FlatMapFunction<M, OM> transformFn, StreamGraphImpl graph, MessageStreamImpl<OM> output) {
-    return new StreamOperatorSpec<>(transformFn, output, OperatorSpec.OpCode.FLAT_MAP, graph.getNextOpId());
+      FlatMapFunction<M, OM> transformFn, MessageStreamImpl<OM> nextStream, int opId) {
+    return new StreamOperatorSpec<>(transformFn, nextStream, OperatorSpec.OpCode.FLAT_MAP, opId);
   }
 
   /**
-   * Creates a {@link SinkOperatorSpec}.
+   * Creates an {@link OutputOperatorSpec} for the sink operator.
    *
-   * @param sinkFn  the sink function
+   * @param sinkFn  the sink function provided by the user
+   * @param opId  the unique ID of the operator
    * @param <M>  type of input message
-   * @param graph  the {@link StreamGraphImpl} object
-   * @return  the {@link SinkOperatorSpec}
+   * @return  the {@link OutputOperatorSpec} for the sink operator
```

Comment:
> It seems that you have renamed the SinkOperatorSpec to OutputOperatorSpec w/o changing the opCode correspondingly. Any particular reason? 

As earlier, the opCodes are passed by OperatorSpecs when it creates the OutputOperatorSpec. So an OutputOperatorSpec opCode can be a SEND_TO, SINK or PARTITION_BY. This change just renames the class to be more general (sink vs output).

> I thought that InputOperatorSpec/OutputOperatorSpec is not in the scope of this change?

Yup, not related to the Input/OutputOperatorSpec in the previous discussion. I couldn't find a consistent way to introduce both, this is as close as I could get.

> Following the line of thought on sink() or sendTo() methods, here are the few things that we had talked about in another PR #51 
> 1) if sink() is only send to non-stream terminals, maybe we should define SinkFunction as SinkFunction(M message, TaskCoordinator coordinator), removing the MessageCollector parameter there?
> 2) For sendTo(), if we implemented SAMZA-1101, it would be translated into an OutputOperatorSpec, which only contains StreamSpec, keyExtractor, msgExtractor, instead of an actual SinkFunction?

Agree with both. I think this PR still achieves the goal of SAMZA-1101 which was to not have the SinkFn for intermediate streams be created in the StreamGraphImpl, and move it down to the OperatorSpec/OperatorImpl layer instead. When removing the MessageCollector from the SinkFn, it's quite likely that we'll end up with 2 different specs - one for sink and one for sendTo and partitionBy. In that version, as you suggest, we'll probably end up not reusing SinkFn for sendTo and partitionBy. 

I think we're mostly on the same page in terms of what we want, I'm just leaving the sink/sendTo split for a later PR. The SinkOperatorSpec -> OutputOperatorSpec renaming is probably causing some confusion, so I can revert the rename and do it in the subsequent PR if that'll be clearer.